File: ureka-framework-main/ureka_framework/logic/stage_worker/msg_generator_r_ticket.py
# Data Model (RAM)
import copy
from pydantic import ValidationError
from ureka_framework.model.data_model.this_device import ThisDevice
from ureka_framework.model.data_model.this_person import ThisPerson
from ureka_framework.model.data_model.other_device import OtherDevice
from ureka_framework.model.message_model.r_ticket import RTicket, r_ticket_to_jsonstr
import ureka_framework.model.message_model.r_ticket as r_ticket
import ureka_framework.model.message_model.u_ticket as u_ticket
from ureka_framework.resource.crypto.serialization_util import (
    str_to_byte,
    byte_to_base64str,
)

# Resource (Crypto)
import ureka_framework.resource.crypto.ecc as ecc
from cryptography.hazmat.primitives.asymmetric import ec
from ureka_framework.resource.crypto import ecdh

# Resource (Logger)
from ureka_framework.resource.logger.simple_logger import simple_log

# Resource (Measurer)
from ureka_framework.resource.logger.simple_measurer import measure_worker_func


class RTicketGenerator:
    def __init__(
        self,
        this_device: ThisDevice,
        this_person: ThisPerson,
        device_table: dict[str, OtherDevice],
    ) -> None:
        self.this_device = this_device
        self.this_person = this_person
        self.device_table = device_table

    ######################################################
    # Message Generation Flow
    ######################################################
    def generate_arbitrary_r_ticket(self, arbitrary_dict: dict) -> RTicket:
        success_msg = "-> SUCCESS: GENERATE_RTICKET"
        failure_msg = "-> FAILURE: GENERATE_RTICKET"

        ######################################################
        # Unsigned RTicket
        ######################################################
        try:
            new_r_ticket = RTicket(**arbitrary_dict)
        except ValidationError as error:  # pragma: no cover -> Weird Ticket-Request
            simple_log("error", f"{failure_msg}: {error}")
            raise RuntimeError(failure_msg)

        # "device"
        if (
            new_r_ticket.r_ticket_type == u_ticket.TYPE_INITIALIZATION_UTICKET
            or new_r_ticket.r_ticket_type == u_ticket.TYPE_OWNERSHIP_UTICKET
            or new_r_ticket.r_ticket_type == r_ticket.TYPE_CRKE1_RTICKET
            or new_r_ticket.r_ticket_type == r_ticket.TYPE_CRKE3_RTICKET
            or new_r_ticket.r_ticket_type == r_ticket.TYPE_DATA_RTOKEN
            or new_r_ticket.r_ticket_type == u_ticket.TYPE_ACCESS_END_UTOKEN
        ):
            new_r_ticket.ticket_order = self.this_device.ticket_order
        # "holder"
        elif new_r_ticket.r_ticket_type == r_ticket.TYPE_CRKE2_RTICKET:
            new_r_ticket.ticket_order = self.device_table[
                new_r_ticket.device_id
            ].ticket_order
        else:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        # Generate UTicket Id (Hash-based)
        new_r_ticket.r_ticket_id = ecdh.generate_sha256_hash_str(
            r_ticket_to_jsonstr(new_r_ticket)
        )

        ######################################################
        # Signed RTicket
        ######################################################
        # Generate Signature
        # "device"
        if (
            new_r_ticket.r_ticket_type == u_ticket.TYPE_INITIALIZATION_UTICKET
            or new_r_ticket.r_ticket_type == u_ticket.TYPE_OWNERSHIP_UTICKET
            or new_r_ticket.r_ticket_type == r_ticket.TYPE_CRKE1_RTICKET
            or new_r_ticket.r_ticket_type == r_ticket.TYPE_CRKE3_RTICKET
            or new_r_ticket.r_ticket_type == u_ticket.TYPE_ACCESS_END_UTOKEN
        ):
            new_r_ticket = self._add_device_signature_on_r_ticket(
                new_r_ticket, self.this_device.device_priv_key
            )
            simple_log("info", success_msg)
        # "holder"
        elif new_r_ticket.r_ticket_type == r_ticket.TYPE_CRKE2_RTICKET:
            new_r_ticket = self._add_device_signature_on_r_ticket(
                new_r_ticket, self.this_person.person_priv_key
            )
            simple_log("info", success_msg)
        # "device"
        elif new_r_ticket.r_ticket_type == r_ticket.TYPE_DATA_RTOKEN:
            # NO Signature
            simple_log("info", success_msg)
        else:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        return new_r_ticket

    ######################################################
    # Add ECC Signature on RTicket
    ######################################################
    def _add_device_signature_on_r_ticket(
        self, unsigned_r_ticket: RTicket, private_key: ec.EllipticCurvePrivateKey
    ) -> RTicket:
        # Message
        unsigned_r_ticket_str = r_ticket_to_jsonstr(unsigned_r_ticket)
        unsigned_r_ticket_byte = str_to_byte(unsigned_r_ticket_str)

        # Sign Signature
        signature_byte = ecc.sign_signature(unsigned_r_ticket_byte, private_key)

        # Add Signature on New Signed RTicket, but Prevent side effect on Unsigned RTicket
        signed_r_ticket = copy.deepcopy(unsigned_r_ticket)
        signed_r_ticket.device_signature = byte_to_base64str(signature_byte)

        return signed_r_ticket

File: ureka-framework-java/app/src/main/java/ureka/framework/logic/pipeline_flow/stage_worker/MsgGeneratorRTicket.java
package ureka.framework.logic.pipeline_flow.stage_worker;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.SignatureException;
import java.security.interfaces.ECPrivateKey;
import java.util.Map;
import java.util.Objects;

import ureka.framework.model.data_model.OtherDevice;
import ureka.framework.model.data_model.ThisDevice;
import ureka.framework.model.data_model.ThisPerson;
import ureka.framework.model.message_model.RTicket;
import ureka.framework.model.message_model.UTicket;
import ureka.framework.resource.crypto.ECC;
import ureka.framework.resource.crypto.ECDH;
import ureka.framework.resource.crypto.SerializationUtil;
import ureka.framework.resource.logger.SimpleLogger;

public class MsgGeneratorRTicket {
    private ThisDevice thisDevice;
    private ThisPerson thisPerson;
    private Map<String, OtherDevice> deviceTable;
    public MsgGeneratorRTicket(ThisDevice thisDevice, ThisPerson thisPerson, Map<String, OtherDevice> deviceTable) {
        this.thisDevice = thisDevice;
        this.thisPerson = thisPerson;
        this.deviceTable = deviceTable;
    }

    // Message Generation Flow
    public RTicket generateArbitraryRTicket(Map<String, String> arbitraryDict) {
        String successMsg = "-> SUCCESS: GENERATE_RTICKET";
        String failureMsg = "-> FAILURE: GENERATE_RTICKET";

        // Unsigned RTicket
        RTicket newRTicket;
        try {
            newRTicket = new RTicket(arbitraryDict);  // something to fixed
        } catch (Exception e) {
            // SimpleLogger.simpleLog("error", failureMsg + ": " + e.getMessage());
            throw new RuntimeException(failureMsg);
        }

        // "device"
        if(Objects.equals(newRTicket.getRTicketType(), UTicket.TYPE_INITIALIZATION_UTICKET) ||
                Objects.equals(newRTicket.getRTicketType(), UTicket.TYPE_OWNERSHIP_UTICKET) ||
                Objects.equals(newRTicket.getRTicketType(), RTicket.TYPE_CRKE1_RTICKET) ||
                Objects.equals(newRTicket.getRTicketType(), RTicket.TYPE_CRKE3_RTICKET) ||
                Objects.equals(newRTicket.getRTicketType(), RTicket.TYPE_DATA_RTOKEN) ||
                Objects.equals(newRTicket.getRTicketType(), UTicket.TYPE_ACCESS_END_UTOKEN)) {
            newRTicket.setTicketOrder(this.thisDevice.getTicketOrder());
        }
        // "holder"
        else if (newRTicket.getRTicketType() == RTicket.TYPE_CRKE2_RTICKET) {
            newRTicket.setTicketOrder(this.deviceTable.get(newRTicket.getDeviceId()).getTicketOrder());
        }
        // pragma: no cover -> Shouldn't Reach Here
        else {
            throw new RuntimeException("Shouldn't Reach Here");
        }

        // Generate RTicket Id (Hash-based)
        try {
            newRTicket.setRTicketId(ECDH.generateSha256HashStr(RTicket.rTicketToJsonStr(newRTicket))); //something to fixed
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        // Signed RTicket
        // Generate Signature
        // "device"
        if(Objects.equals(newRTicket.getRTicketType(), UTicket.TYPE_INITIALIZATION_UTICKET) ||
                Objects.equals(newRTicket.getRTicketType(), UTicket.TYPE_OWNERSHIP_UTICKET) ||
                Objects.equals(newRTicket.getRTicketType(), RTicket.TYPE_CRKE1_RTICKET) ||
                Objects.equals(newRTicket.getRTicketType(), RTicket.TYPE_CRKE3_RTICKET) ||
                Objects.equals(newRTicket.getRTicketType(), UTicket.TYPE_ACCESS_END_UTOKEN)) {
            newRTicket = _addIssuerSignatureOnRTicket(newRTicket,this.thisDevice.getDevicePrivKey());
            SimpleLogger.simpleLog("info", successMsg);
        }
        // "holder"
        else if (Objects.equals(newRTicket.getRTicketType(), RTicket.TYPE_CRKE2_RTICKET)) {
            newRTicket = _addIssuerSignatureOnRTicket(newRTicket,this.thisPerson.getPersonPrivKey());
            SimpleLogger.simpleLog("info", successMsg);
        }
        // "device"
        else if (Objects.equals(newRTicket.getRTicketType(), RTicket.TYPE_DATA_RTOKEN)) {
            // NO Signature
            SimpleLogger.simpleLog("info", successMsg);
        }
        // pragma: no cover -> Shouldn't Reach Here
        else {
            throw new RuntimeException("Shouldn't Reach Here");
        }

        return newRTicket;
    }

    // something to fixed
    // Add ECC Signature on RTicket
    private RTicket _addIssuerSignatureOnRTicket(RTicket unsignedRTicket, ECPrivateKey privateKey) {
        // Message
        String unsignedRTicketStr = RTicket.rTicketToJsonStr(unsignedRTicket);  // something to fixed
        byte[] unsignedRTicketByte = SerializationUtil.strToByte(unsignedRTicketStr);

        // Sign Signature
        byte[] signatureByte;
        try {
            signatureByte = ECC.signSignature(unsignedRTicketByte, privateKey);
        } catch (NoSuchAlgorithmException | NoSuchProviderException | InvalidKeyException |
                 SignatureException e) {
            throw new RuntimeException(e);
        }

        // Add Signature on New Signed RTicket, but Prevent side effect on Unsigned RTicket
        RTicket signedRTicket;
        try {
            signedRTicket = deepCopy(unsignedRTicket);
        } catch (IOException | ClassNotFoundException e) {
            throw new RuntimeException(e);
        }
        signedRTicket.setDeviceSignature(SerializationUtil.byteToBase64Str(signatureByte));

        return signedRTicket;
    }

    private RTicket deepCopy(RTicket originalRTicket) throws IOException, ClassNotFoundException {
        // Serialize to a byte array
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        ObjectOutputStream oos = new ObjectOutputStream(bos);
        oos.writeObject(originalRTicket);
        oos.flush();
        oos.close();
        bos.close();

        // Deserialize from byte array
        ByteArrayInputStream bis = new ByteArrayInputStream(bos.toByteArray());
        ObjectInputStream ois = new ObjectInputStream(bis);
        RTicket copyRTicket = (RTicket) ois.readObject();
        ois.close();
        bis.close();

        return copyRTicket;
    }
}
